### DISCUSSION TITLE
v2版本的DatePicker日期选择组件选中区域高度在某些机型上的高度bug及解决方案

### DISCUSSION THREAD

**Discussion by damonyoungcc** [ant-design/ant-design-mobile#5233] [Q&A] [answered]
### Version of antd-mobile

v2

### Operating system and its version

Android

### What happened?

如下图，在大多数机型中都可以正常显示，但是在某些机型中（mate 40 鸿蒙系统 或 某些小米机型），选中高亮区域高度会变得很高
![02](https://user-images.githubusercontent.com/15868859/170171452-c6a8e2d7-72da-4767-9ede-bd8b24e85d5e.jpeg)
![01](https://user-images.githubusercontent.com/15868859/170171972-6e3ea6e1-0e74-415d-90b8-96715857907a.jpeg)

**Comment by awmleer** [comment]
要不转成 discussion 吧？issue close 了就不好找到了 😵

**Comment by damonyoungcc** [comment, marked as answer]
## 原因分析

 初步分析，原因是`am-picker-col-mask`的`backgroundSize`的高度出现错误导致，如下图

![03](https://user-images.githubusercontent.com/15868859/170172116-8fbb16de-0e63-4e80-ae0f-88aac2822f9a.jpeg)

然后往上追溯源码，发现v2的日期组件最终依赖 -> [rmc-picker](https://github.com/react-component/m-picker)

查看源码后，发现`backgroundSize`行内样式高度的设置代码，在这里 [第178行](https://github.com/react-component/m-picker/blob/master/src/Picker.tsx#L178)，下面是源码
```tsx
    const { contentRef, indicatorRef, maskRef, rootRef } = this;
    const rootHeight = rootRef.getBoundingClientRect().height;
    // https://github.com/react-component/m-picker/issues/18
    const itemHeight = this.itemHeight = indicatorRef.getBoundingClientRect().height;
    let num = Math.floor(rootHeight / itemHeight);
    if (num % 2 === 0) {
      num--;
    }
    num--;
    num /= 2;
    contentRef.style.padding = `${itemHeight * num}px 0`;
    indicatorRef.style.top = `${itemHeight * num}px`;
    maskRef.style.backgroundSize = `100% ${itemHeight * num}px`;
```

经过验证发现，`getBoundingClientRect`获取的高度，在某些机型上可能会是浮点数，因此导致num计算出现错误，从而导致设置`mask`的`backgroundSize`错误。
在出问题的机型上打印结果，果然如此，如下图
![04](https://user-images.githubusercontent.com/15868859/170173474-1f68d64e-27c1-497c-9485-18d1ccf5e0f4.jpeg)

**Comment by damonyoungcc** [comment]
> 要不转成 discussion 吧？issue close 了就不好找到了 😵

好的

**Reply by awmleer** [reply]
done 我操作了

**Reply by damonyoungcc** [reply]
## 目前我的解决方案：
在计算`num`的时候，把原先的`Math. floor`改成`Math. round`即可，如果只想在业务端快速修复`bug`，思路是在`DatePicker`的`onVisibleChange`回调中（文档中并没有标出有这个方法，源码中是提供了此方法的，[戳这里](https://github.com/ant-design/ant-design-mobile/blob/v2/components/date-picker/index.tsx#L17) ） 重写计算`num`设置`backgroundSize`等相关的逻辑。

```tsx
const onVisibleChange = (visible) => {
  if (visible) {
    setTimeout(() => {
      const maskElementList = document.querySelectorAll(".am-picker-col-mask");
      const rootElementList = document.querySelectorAll(".am-picker-col");
      const itemElementList = document.querySelectorAll(".am-picker-col-indicator");
      const contentElementList = document.querySelectorAll(".am-picker-col-content");

      const rootHeight = rootElementList[0].getBoundingClientRect().height;
      const itemHeight = itemElementList[0].getBoundingClientRect().height;

      let num = Math.round(rootHeight / itemHeight);
      if (num % 2 === 0) {
        num--;
      }
      num--;
      num /= 2;

      contentElementList.forEach((item: HTMLElement) => {
        item.style.padding = `${itemHeight * num}px 0`;
      });
      itemElementList.forEach((item: HTMLElement) => {
        item.style.top = `${itemHeight * num}px`;
      });
      maskElementList.forEach((item: HTMLElement) => {
        item.style.backgroundSize = `100% ${itemHeight * num}px`;
      });
    });
  }
};
```
